### PR TITLE
Support user-provided font-fallback.

### PR DESCRIPTION
**Pull request by GaryQian**
This adds the `fontFamilyFallback` field to TextStyle and passes it to minikin.

### CODE REVIEW

**Comment by GaryQian** [comment]
Will require a manual engine roll to support the framework-side changes

File: lib/ui/text.dart [hunk 1]
```diff
@@ -340,6 +345,7 @@ class TextStyle {
   /// * `fontWeight`: The typeface thickness to use when painting the text (e.g., bold).
   /// * `fontStyle`: The typeface variant to use when drawing the letters (e.g., italics).
   /// * `fontFamily`: The name of the font to use when painting the text (e.g., Roboto).
+  /// * `fontFamily`: An ordered list of the names of the fonts to fallback on when a glyph cannot be found in a higher priority font.
```
Comment (jason-simmons) [inline lib/ui/text.dart]: should be `fontFamilyFallback`

File: lib/ui/text.dart [hunk 2]
```diff
@@ -1135,11 +1164,25 @@ class ParagraphBuilder extends NativeFieldWrapperClass2 {
...
+  void pushStyle(TextStyle style) => _pushStyle(style._encoded, style._fontFamily, _encodeFontFamilyFallback(style._fontFamilyFallback), style._fontSize, style._letterSpacing, style._wordSpacing, style._height, _encodeLocale(style._locale), style._background?._objects, style._background?._data, style._foreground?._objects, style._foreground?._data, Shadow._encodeShadows(style._shadows));
+  void _pushStyle(Int32List encoded, String fontFamily, String fontFamilyFallback, double fontSize, double letterSpacing, double wordSpacing, double height, String locale, List<dynamic> backgroundObjects, ByteData backgroundData, List<dynamic> foregroundObjects, ByteData foregroundData, ByteData shadowsData) native 'ParagraphBuilder_pushStyle';
 
   static String _encodeLocale(Locale locale) => locale?.toString() ?? '';
 
+  // Generates a comma-separated string that includes all of the fallback font families.
+  static String _encodeFontFamilyFallback(List<String> fontFamilyFallback) {
```
Comment (jason-simmons) [inline lib/ui/text.dart]: Tonic has a converter for `std::vector`.  Does the conversion to `vector<string>` work if you pass the fallback list as a `List<String>` in `_pushStyle`?

File: third_party/txt/src/txt/text_style.cc [hunk 3]
```diff
@@ -53,6 +53,11 @@ bool TextStyle::equals(const TextStyle& other) const {
     return false;
   if (text_shadows.size() != other.text_shadows.size())
     return false;
+  for (size_t font_index = 0; font_index < font_family_fallback.size();
```
Comment (jason-simmons) [inline third_party/txt/src/txt/text_style.cc]: Add a check that `font_family_fallback.size()` matches `other.font_family_fallback.size()`

Also, move this so it is not between the checks for `text_shadows`

File: lib/ui/text/paragraph_builder.cc [hunk 4]
```diff
@@ -253,7 +256,8 @@ void ParagraphBuilder::pushStyle(tonic::Int32List& encoded,
       style.font_style = static_cast<txt::FontStyle>(encoded[tsFontStyleIndex]);
 
     if (mask & tsFontFamilyMask)
-      style.font_family = fontFamily;
+      style.font_families.push_back(fontFamily);// = fontFamily;
```
Comment (jason-simmons) [inline lib/ui/text/paragraph_builder.cc]: Remove these comments

(If fontFamily is merged into fontFamilyFallback, then this block will no longer be needed)

File: lib/ui/text.dart [hunk 5]
```diff
@@ -340,6 +345,7 @@ class TextStyle {
   /// * `fontWeight`: The typeface thickness to use when painting the text (e.g., bold).
   /// * `fontStyle`: The typeface variant to use when drawing the letters (e.g., italics).
   /// * `fontFamily`: The name of the font to use when painting the text (e.g., Roboto).
+  /// * `fontFamilyFallback`: An ordered list of the names of the fonts to fallback on when a glyph cannot be found in a higher priority font.
```
Comment (Hixie) [inline lib/ui/text.dart]: can the list be null? can it be empty?

File: lib/ui/text.dart [hunk 6]
```diff
@@ -428,33 +438,52 @@ class TextStyle {
...
       return false;
     return true;
   }
 
+  /// Determines if lists [a] and [b] are deep equivalent.
+  ///
+  /// Returns true if the lists are both null, or if they are both non-null, have
+  /// the same length, and contain the same elements in the same order. Returns
+  /// false otherwise.
+  static bool _listEquals(List<dynamic> a, List<dynamic> b) {
```
Comment (Hixie) [inline lib/ui/text.dart]: `_listEquals<T>(List<T> a, List<T> b) ...`

File: third_party/txt/src/txt/font_collection.cc [hunk 7]
```diff
@@ -112,57 +123,77 @@ void FontCollection::DisableFontFallback() {
...
+  size_t font_families_count = font_families.size();
+
+  // Attempt to match each font family in order.
+  for (size_t fallback_index = 0; fallback_index < font_families_count + 1;
+       fallback_index++) {
+    std::string font_family;
+    // When we run out of fonts to attempt to find, try to match the default
+    // font.
+    if (fallback_index == font_families_count) {
+      font_family = default_font_family;
```
Comment (jason-simmons) [inline third_party/txt/src/txt/font_collection.cc]: Previously the engine would try the style's font family in all font managers and then use the default font family if the style's font could not be matched anywhere.

Can we keep that prioritization?  For each family in the fallback list, try it in each font manager.  Only resort to the default font family if none of those combinations work.

File: lib/ui/text.dart (hunk 6 above)
Comment (Hixie) [reply lib/ui/text.dart]: (see https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/collections.dart#L28)

File: lib/ui/text/paragraph_builder.cc [hunk 8]
```diff
@@ -253,7 +256,8 @@ void ParagraphBuilder::pushStyle(tonic::Int32List& encoded,
...
+      // style.font_family = fontFamily;
```
Comment (Hixie) [inline lib/ui/text/paragraph_builder.cc]: commented-out code

**Comment by Hixie** [comment]
Is there any reason to pass the fontFamily and the fontFamilyFallback as separate items from the dart to the C++? It seems like we could just pass one list, which is fontFamily + fontFamilyFallback, and the C++ code wouldn't have to know anything about the distinction.

File: lib/ui/text.dart (hunk 5 above)
Comment (Hixie) [reply lib/ui/text.dart]: can it be specified if fontFamily is null?

File: third_party/txt/src/txt/font_collection.cc (hunk 7 above)
Comment (GaryQian) [reply third_party/txt/src/txt/font_collection.cc]: This change should preserve that behavior. The `fallback_index` increases to font_families.size() + 1, where the extra iteration is caught here and set to the default font at the end when every other option has been exhausted.

**Pull request merged**